File: go/graphics/mp3_wav_converter.go
// MP3-to-Wav Converter — audio buffer + resample + normalize + RIFF encoding.
package graphics

import (
	"encoding/binary"
	"fmt"
)

type AudioBuffer struct {
	SampleRate uint32
	Channels   uint16
	Samples    []int16
}

func (b *AudioBuffer) FrameCount() int {
	if b.Channels == 0 {
		return 0
	}
	return len(b.Samples) / int(b.Channels)
}

func (b *AudioBuffer) DurationMs() uint64 {
	if b.SampleRate == 0 {
		return 0
	}
	return uint64(b.FrameCount()) * 1000 / uint64(b.SampleRate)
}

type Mp3Frame struct {
	SampleRate       uint32
	Channels         uint16
	BitrateKbps      uint32
	SamplesPerFrame  uint32
	Pcm              []int16
}

type CodecError struct{ Reason string }

func (e *CodecError) Error() string { return e.Reason }

func DecodeMp3ToBuffer(frames []Mp3Frame) (*AudioBuffer, error) {
	if len(frames) == 0 {
		return nil, &CodecError{"no_frames"}
	}
	sr := frames[0].SampleRate
	ch := frames[0].Channels
	if sr == 0 {
		return nil, &CodecError{"sample_rate_zero"}
	}
	if ch != 1 && ch != 2 {
		return nil, &CodecError{fmt.Sprintf("unsupported_channels: %d", ch)}
	}
	for _, f := range frames {
		if f.SampleRate != sr || f.Channels != ch {
			return nil, &CodecError{"mixed_formats"}
		}
	}
	var samples []int16
	for _, f := range frames {
		samples = append(samples, f.Pcm...)
	}
	return &AudioBuffer{SampleRate: sr, Channels: ch, Samples: samples}, nil
}

func clampI16(v int64) int16 {
	if v > 32767 {
		return 32767
	}
	if v < -32768 {
		return -32768
	}
	return int16(v)
}

func Resample(buf *AudioBuffer, dstRate uint32) (*AudioBuffer, error) {
	if buf.SampleRate == 0 || dstRate == 0 {
		return nil, &CodecError{"sample_rate_zero"}
	}
	if dstRate == buf.SampleRate {
		out := make([]int16, len(buf.Samples))
		copy(out, buf.Samples)
		return &AudioBuffer{SampleRate: buf.SampleRate, Channels: buf.Channels, Samples: out}, nil
	}
	inFrames := buf.FrameCount()
	if inFrames == 0 {
		return &AudioBuffer{SampleRate: dstRate, Channels: buf.Channels}, nil
	}
	outFrames := (uint64(inFrames)*uint64(dstRate) + uint64(buf.SampleRate)/2) / uint64(buf.SampleRate)
	channels := int(buf.Channels)
	out := make([]int16, 0, int(outFrames)*channels)
	for i := uint64(0); i < outFrames; i++ {
		srcPos := i * uint64(buf.SampleRate) * 1000 / uint64(dstRate)
		srcFrame := int(srcPos / 1000)
		frac := int64(srcPos % 1000)
		nextFrame := srcFrame + 1
		if nextFrame >= inFrames {
			nextFrame = inFrames - 1
		}
		for c := 0; c < channels; c++ {
			a := int64(buf.Samples[srcFrame*channels+c])
			b := int64(buf.Samples[nextFrame*channels+c])
			v := a + (b-a)*frac/1000
			out = append(out, clampI16(v))
		}
	}
	return &AudioBuffer{SampleRate: dstRate, Channels: buf.Channels, Samples: out}, nil
}

func NormalizeToPeak(buf *AudioBuffer, targetAmplitude int32) *AudioBuffer {
	if len(buf.Samples) == 0 {
		return &AudioBuffer{SampleRate: buf.SampleRate, Channels: buf.Channels}
	}
	var maxAbs int32
	for _, s := range buf.Samples {
		a := int32(s)
		if a < 0 {
			a = -a
		}
		if a > maxAbs {
			maxAbs = a
		}
	}
	if maxAbs == 0 {
		out := make([]int16, len(buf.Samples))
		copy(out, buf.Samples)
		return &AudioBuffer{SampleRate: buf.SampleRate, Channels: buf.Channels, Samples: out}
	}
	target := targetAmplitude
	if target > 32767 {
		target = 32767
	}
	if target < 0 {
		target = 0
	}
	out := make([]int16, len(buf.Samples))
	for i, s := range buf.Samples {
		v := int64(s) * int64(target) / int64(maxAbs)
		out[i] = clampI16(v)
	}
	return &AudioBuffer{SampleRate: buf.SampleRate, Channels: buf.Channels, Samples: out}
}

func ToMono(buf *AudioBuffer) *AudioBuffer {
	if buf.Channels == 1 {
		out := make([]int16, len(buf.Samples))
		copy(out, buf.Samples)
		return &AudioBuffer{SampleRate: buf.SampleRate, Channels: 1, Samples: out}
	}
	n := buf.FrameCount()
	out := make([]int16, n)
	for i := 0; i < n; i++ {
		l := int32(buf.Samples[i*2])
		r := int32(buf.Samples[i*2+1])
		out[i] = int16((l + r) / 2)
	}
	return &AudioBuffer{SampleRate: buf.SampleRate, Channels: 1, Samples: out}
}

func ToStereo(buf *AudioBuffer) *AudioBuffer {
	if buf.Channels == 2 {
		out := make([]int16, len(buf.Samples))
		copy(out, buf.Samples)
		return &AudioBuffer{SampleRate: buf.SampleRate, Channels: 2, Samples: out}
	}
	out := make([]int16, 0, len(buf.Samples)*2)
	for _, s := range buf.Samples {
		out = append(out, s, s)
	}
	return &AudioBuffer{SampleRate: buf.SampleRate, Channels: 2, Samples: out}
}

func EncodeWav(buf *AudioBuffer) []byte {
	bitsPerSample := uint16(16)
	byteRate := buf.SampleRate * uint32(buf.Channels) * uint32(bitsPerSample/8)
	blockAlign := buf.Channels * (bitsPerSample / 8)
	dataSize := uint32(len(buf.Samples) * 2)
	riffSize := 36 + dataSize

	out := make([]byte, 0, 44+int(dataSize))
	out = append(out, "RIFF"...)
	out = binary.LittleEndian.AppendUint32(out, riffSize)
	out = append(out, "WAVE"...)
	out = append(out, "fmt "...)
	out = binary.LittleEndian.AppendUint32(out, 16)
	out = binary.LittleEndian.AppendUint16(out, 1)
	out = binary.LittleEndian.AppendUint16(out, buf.Channels)
	out = binary.LittleEndian.AppendUint32(out, buf.SampleRate)
	out = binary.LittleEndian.AppendUint32(out, byteRate)
	out = binary.LittleEndian.AppendUint16(out, blockAlign)
	out = binary.LittleEndian.AppendUint16(out, bitsPerSample)
	out = append(out, "data"...)
	out = binary.LittleEndian.AppendUint32(out, dataSize)
	for _, s := range buf.Samples {
		out = binary.LittleEndian.AppendUint16(out, uint16(s))
	}
	return out
}

func DecodeWav(data []byte) (*AudioBuffer, error) {
	if len(data) < 44 {
		return nil, &CodecError{"empty_buffer"}
	}
	if string(data[0:4]) != "RIFF" || string(data[8:12]) != "WAVE" {
		return nil, &CodecError{"empty_buffer"}
	}
	channels := binary.LittleEndian.Uint16(data[22:24])
	sampleRate := binary.LittleEndian.Uint32(data[24:28])
	dataSize := int(binary.LittleEndian.Uint32(data[40:44]))
	samples := make([]int16, dataSize/2)
	for i := 0; i < dataSize/2; i++ {
		samples[i] = int16(binary.LittleEndian.Uint16(data[44+i*2 : 44+i*2+2]))
	}
	return &AudioBuffer{SampleRate: sampleRate, Channels: channels, Samples: samples}, nil
}

func Mp3WavConverterDemo() {
	frames := []Mp3Frame{
		{SampleRate: 22050, Channels: 2, BitrateKbps: 128, SamplesPerFrame: 4,
			Pcm: []int16{100, -100, 200, -200, 400, -400, 800, -800}},
		{SampleRate: 22050, Channels: 2, BitrateKbps: 128, SamplesPerFrame: 2,
			Pcm: []int16{-500, 500, -1000, 1000}},
	}
	buf, _ := DecodeMp3ToBuffer(frames)
	fmt.Printf("decoded: rate=%d ch=%d samples=%d\n", buf.SampleRate, buf.Channels, len(buf.Samples))

	resampled, _ := Resample(buf, 44100)
	fmt.Printf("resampled: rate=%d ch=%d samples=%d\n", resampled.SampleRate, resampled.Channels, len(resampled.Samples))

	normalized := NormalizeToPeak(resampled, 32000)
	var maxAbs int32
	for _, s := range normalized.Samples {
		a := int32(s)
		if a < 0 { a = -a }
		if a > maxAbs { maxAbs = a }
	}
	fmt.Printf("normalized max_abs: %d\n", maxAbs)

	mono := ToMono(normalized)
	fmt.Printf("mono: ch=%d samples=%d\n", mono.Channels, len(mono.Samples))

	wav := EncodeWav(mono)
	fmt.Printf("wav bytes: %d\n", len(wav))
	fmt.Printf("header: %s %s\n", wav[0:4], wav[8:12])

	decoded, _ := DecodeWav(wav)
	match := len(decoded.Samples) == len(mono.Samples)
	if match {
		for i := range decoded.Samples {
			if decoded.Samples[i] != mono.Samples[i] {
				match = false
				break
			}
		}
	}
	fmt.Printf("decoded wav: rate=%d ch=%d samples=%d\n", decoded.SampleRate, decoded.Channels, len(decoded.Samples))
	fmt.Printf("samples match: %v\n", match)
}

File: python/graphics/mp3_wav_converter.py
"""MP3-to-Wav Converter — audio buffer + resample + normalize + RIFF encoding.

AudioBuffer: interleaved PCM int16 + sample_rate + channels.
Resample: linear interpolation, integer-only math for determinism.
Normalize: two-pass scan max_abs, then scale to target peak.
Channel mix: stereo/mono via average / duplicate.
RIFF encoding: canonical 44-byte WAV header + little-endian samples.
"""

from __future__ import annotations
import struct
from dataclasses import dataclass, field
from enum import Enum


@dataclass
class AudioBuffer:
    sample_rate: int
    channels: int
    samples: list[int] = field(default_factory=list)

    def frame_count(self) -> int:
        if self.channels == 0: return 0
        return len(self.samples) // self.channels

    def duration_ms(self) -> int:
        if self.sample_rate == 0: return 0
        return self.frame_count() * 1000 // self.sample_rate


@dataclass
class Mp3Frame:
    sample_rate: int
    channels: int
    bitrate_kbps: int
    samples_per_frame: int
    pcm: list[int]


class CodecError(Exception):
    pass


class NoFrames(CodecError):
    pass


class MixedFormats(CodecError):
    pass


class EmptyBuffer(CodecError):
    pass


class UnsupportedChannels(CodecError):
    def __init__(self, n: int):
        super().__init__(f"unsupported channels: {n}")
        self.n = n


class SampleRateZero(CodecError):
    pass


def decode_mp3_to_buffer(frames: list[Mp3Frame]) -> AudioBuffer:
    if not frames:
        raise NoFrames()
    sample_rate = frames[0].sample_rate
    channels = frames[0].channels
    if sample_rate == 0:
        raise SampleRateZero()
    if channels not in (1, 2):
        raise UnsupportedChannels(channels)
    for f in frames:
        if f.sample_rate != sample_rate or f.channels != channels:
            raise MixedFormats()
    samples: list[int] = []
    for f in frames:
        samples.extend(f.pcm)
    return AudioBuffer(sample_rate=sample_rate, channels=channels, samples=samples)


def _clamp_i16(v: int) -> int:
    if v > 32767: return 32767
    if v < -32768: return -32768
    return v


def resample(buf: AudioBuffer, dst_rate: int) -> AudioBuffer:
    if buf.sample_rate == 0 or dst_rate == 0:
        raise SampleRateZero()
    if dst_rate == buf.sample_rate:
        return AudioBuffer(buf.sample_rate, buf.channels, list(buf.samples))
    in_frames = buf.frame_count()
    if in_frames == 0:
        return AudioBuffer(dst_rate, buf.channels, [])
    out_frames = (in_frames * dst_rate + buf.sample_rate // 2) // buf.sample_rate
    channels = buf.channels
    out: list[int] = []
    for i in range(out_frames):
        src_pos = i * buf.sample_rate * 1000 // dst_rate
        src_frame = src_pos // 1000
        frac = src_pos % 1000
        next_frame = min(src_frame + 1, in_frames - 1)
        for c in range(channels):
            a = buf.samples[src_frame * channels + c]
            b = buf.samples[next_frame * channels + c]
            v = a + (b - a) * frac // 1000
            out.append(_clamp_i16(v))
    return AudioBuffer(dst_rate, channels, out)


def normalize_to_peak(buf: AudioBuffer, target_amplitude: int) -> AudioBuffer:
    if not buf.samples:
        return AudioBuffer(buf.sample_rate, buf.channels, list(buf.samples))
    max_abs = 0
    for s in buf.samples:
        a = abs(s)
        if a > max_abs:
            max_abs = a
    if max_abs == 0:
        return AudioBuffer(buf.sample_rate, buf.channels, list(buf.samples))
    target = max(0, min(32767, target_amplitude))
    scaled = [_clamp_i16(s * target // max_abs) for s in buf.samples]
    return AudioBuffer(buf.sample_rate, buf.channels, scaled)


def to_mono(buf: AudioBuffer) -> AudioBuffer:
    if buf.channels == 1:
        return AudioBuffer(buf.sample_rate, 1, list(buf.samples))
    n = buf.frame_count()
    out = []
    for i in range(n):
        l = buf.samples[i * 2]
        r = buf.samples[i * 2 + 1]
        out.append((l + r) // 2)
    return AudioBuffer(buf.sample_rate, 1, out)


def to_stereo(buf: AudioBuffer) -> AudioBuffer:
    if buf.channels == 2:
        return AudioBuffer(buf.sample_rate, 2, list(buf.samples))
    out = []
    for s in buf.samples:
        out.append(s)
        out.append(s)
    return AudioBuffer(buf.sample_rate, 2, out)


def encode_wav(buf: AudioBuffer) -> bytes:
    bits_per_sample = 16
    byte_rate = buf.sample_rate * buf.channels * (bits_per_sample // 8)
    block_align = buf.channels * (bits_per_sample // 8)
    data_size = len(buf.samples) * 2
    riff_size = 36 + data_size
    header = struct.pack(
        "<4sI4s4sIHHIIHH4sI",
        b"RIFF", riff_size,
        b"WAVE",
        b"fmt ", 16, 1, buf.channels, buf.sample_rate,
        byte_rate, block_align, bits_per_sample,
        b"data", data_size,
    )
    body = struct.pack(f"<{len(buf.samples)}h", *buf.samples) if buf.samples else b""
    return header + body


def decode_wav(data: bytes) -> AudioBuffer:
    if len(data) < 44:
        raise EmptyBuffer()
    if data[0:4] != b"RIFF" or data[8:12] != b"WAVE":
        raise EmptyBuffer()
    channels = struct.unpack("<H", data[22:24])[0]
    sample_rate = struct.unpack("<I", data[24:28])[0]
    data_size = struct.unpack("<I", data[40:44])[0]
    samples_bytes = data[44:44 + data_size]
    sample_count = len(samples_bytes) // 2
    samples = list(struct.unpack(f"<{sample_count}h", samples_bytes))
    return AudioBuffer(sample_rate, channels, samples)


def demo() -> None:
    frames = [
        Mp3Frame(22050, 2, 128, 4, [100, -100, 200, -200, 400, -400, 800, -800]),
        Mp3Frame(22050, 2, 128, 2, [-500, 500, -1000, 1000]),
    ]
    buf = decode_mp3_to_buffer(frames)
    print(f"decoded: rate={buf.sample_rate} ch={buf.channels} samples={len(buf.samples)}")

    resampled = resample(buf, 44100)
    print(f"resampled: rate={resampled.sample_rate} ch={resampled.channels} samples={len(resampled.samples)}")

    normalized = normalize_to_peak(resampled, 32000)
    max_abs = max(abs(s) for s in normalized.samples)
    print(f"normalized max_abs: {max_abs}")

    mono = to_mono(normalized)
    print(f"mono: ch={mono.channels} samples={len(mono.samples)}")

    wav = encode_wav(mono)
    print(f"wav bytes: {len(wav)}")
    print(f"header: {wav[0:4]!r} {wav[8:12]!r}")

    decoded = decode_wav(wav)
    print(f"decoded wav: rate={decoded.sample_rate} ch={decoded.channels} samples={len(decoded.samples)}")
    print(f"samples match: {decoded.samples == mono.samples}")


if __name__ == "__main__":
    demo()
